File: src/core/classification/boundary-matcher.ts
/**
 * Boundary matcher — strategy-dispatched provider/consumer pairing.
 *
 * Pure core business logic. No I/O, no storage, no framework awareness.
 * The orchestrator dispatches by mechanism to the correct match strategy.
 *
 * Architecture:
 *   - BoundaryMatchStrategy: one per mechanism (http, grpc, ioctl, ...)
 *   - Each strategy owns its own address normalization and matching rules.
 *   - The orchestrator groups facts by mechanism and delegates matching.
 *   - computeMatcherKey: used by the indexer at persist time to populate
 *     the matcher_key column. Same normalization as the match logic.
 *
 * HTTP matching:
 *   Structured segment normalization. Path is split into segments.
 *   Literal segments compare exactly. Parameter segments ({id}, :id,
 *   {param}) normalize to wildcard token {_}. Method is prefixed.
 *   Matching is by normalized key equality.
 *
 * Maturity: PROTOTYPE. Sufficient to validate the boundary-fact model.
 * Known limitations:
 *   - No query parameter matching
 *   - No route prefix/wildcard segment matching (e.g. /api/**)
 *   - No cross-mechanism matching (HTTP consumer → gRPC provider)
 *   - Confidence scoring is a simple product, not calibrated
 */

import type {
	BoundaryConsumerFact,
	BoundaryLinkCandidate,
	BoundaryMechanism,
	BoundaryProviderFact,
} from "./api-boundary.js";

// ── Matchable fact types ────────────────────────────────────────────

/**
 * A provider fact with a stable persisted UID.
 * The matcher requires UIDs so that link candidates carry stable
 * identifiers back to the persisted facts — no object-identity
 * assumptions across the strategy boundary.
 */
export type MatchableProviderFact = BoundaryProviderFact & {
	factUid: string;
};

/**
 * A consumer fact with a stable persisted UID.
 */
export type MatchableConsumerFact = BoundaryConsumerFact & {
	factUid: string;
};

// ── Strategy interface ──────────────────────────────────────────────

/**
 * A mechanism-specific match strategy. One per boundary mechanism.
 *
 * Responsibilities:
 *   1. Normalize an address into a matcher key string
 *   2. Match providers to consumers for this mechanism
 *
 * Contract: the match method receives facts with stable persisted
 * UIDs. Link candidates MUST carry those UIDs, not object references.
 * Strategies are free to normalize, clone, or reconstruct working
 * objects internally — the UID is the stable association.
 */
export interface BoundaryMatchStrategy {
	/** The mechanism this strategy handles. */
	readonly mechanism: BoundaryMechanism;

	/**
	 * Compute the normalized matcher key for an address.
	 *
	 * Used at two sites:
	 *   - Persist time (indexer computes key for the DB column)
	 *   - Match time (strategy normalizes before comparing)
	 *
	 * @param address - The raw address from the fact (e.g. "/api/v2/products/{id}")
	 * @param metadata - Mechanism-specific metadata (e.g. { httpMethod: "GET" })
	 */
	computeMatcherKey(address: string, metadata: Record<string, unknown>): string;

	/**
	 * Match providers against consumers for this mechanism.
	 * Both arrays are pre-filtered to this mechanism.
	 * Returns zero or more link candidates carrying persisted fact UIDs.
	 */
	match(
		providers: MatchableProviderFact[],
		consumers: MatchableConsumerFact[],
	): BoundaryLinkCandidate[];
}

// ── HTTP match strategy ─────────────────────────────────────────────

/** Wildcard token used in normalized path segments for any parameter. */
const PARAM_WILDCARD = "{_}";

/**
 * Normalize a single path segment for matching.
 *
 * Patterns recognized as parameters:
 *   - {id}, {productId}, {param} — curly-brace path variable (Spring, consumer extractor)
 *   - :id, :productId — colon path parameter (Express, Koa)
 *
 * Everything else is treated as a literal segment.
 */
function normalizeSegment(segment: string): string {
	if (segment.startsWith("{") && segment.endsWith("}")) return PARAM_WILDCARD;
	if (segment.startsWith(":") && segment.length > 1) return PARAM_WILDCARD;
	return segment;
}

/**
 * Normalize an HTTP path into a canonical segment representation.
 *
 * Input: "/api/v2/products/{id}/reviews"
 * Output: "/api/v2/products/{_}/reviews"
 *
 * The raw path is preserved in the fact; this produces only the
 * matching key. Parameter names are discarded for matching purposes
 * because provider and consumer use different naming conventions
 * (Spring {id} vs consumer {param} vs Express :productId).
 */
function normalizeHttpPath(path: string): string {
	// Handle empty or missing path.
	if (!path || path === "/") return "/";

	const segments = path.split("/");
	return segments.map(normalizeSegment).join("/");
}

/**
 * Count how many segments in a normalized path are wildcards vs literals.
 * Used for confidence adjustment — more literal segments = higher confidence.
 */
function countSegmentTypes(normalizedPath: string): {
	total: number;
	literal: number;
	wildcard: number;
} {
	const segments = normalizedPath.split("/").filter((s) => s.length > 0);
	let wildcard = 0;
	for (const s of segments) {
		if (s === PARAM_WILDCARD) wildcard++;
	}
	return {
		total: segments.length,
		literal: segments.length - wildcard,
		wildcard,
	};
}

export class HttpBoundaryMatchStrategy implements BoundaryMatchStrategy {
	readonly mechanism: BoundaryMechanism = "http";

	computeMatcherKey(
		address: string,
		metadata: Record<string, unknown>,
	): string {
		const method =
			typeof metadata.httpMethod === "string"
				? metadata.httpMethod.toUpperCase()
				: "*";
		const normalizedPath = normalizeHttpPath(address);
		return `${method} ${normalizedPath}`;
	}

	match(
		providers: MatchableProviderFact[],
		consumers: MatchableConsumerFact[],
	): BoundaryLinkCandidate[] {
		if (providers.length === 0 || consumers.length === 0) return [];

		// Build provider index: matcherKey → provider facts.
		const providerIndex = new Map<string, MatchableProviderFact[]>();
		// Also build path-only index for consumers without a known method.
		const providerByPathOnly = new Map<string, MatchableProviderFact[]>();

		for (const p of providers) {
			const key = this.computeMatcherKey(p.address, p.metadata);
			if (!providerIndex.has(key)) providerIndex.set(key, []);
			providerIndex.get(key)!.push(p);

			// Path-only key (strip method prefix).
			const pathOnly = normalizeHttpPath(p.address);
			if (!providerByPathOnly.has(pathOnly)) {
				providerByPathOnly.set(pathOnly, []);
			}
			providerByPathOnly.get(pathOnly)!.push(p);
		}

		const candidates: BoundaryLinkCandidate[] = [];

		for (const c of consumers) {
			const consumerKey = this.computeMatcherKey(c.address, c.metadata);
			const method =
				typeof c.metadata.httpMethod === "string"
					? c.metadata.httpMethod.toUpperCase()
					: null;

			// Strategy 1: exact key match (method + normalized path).
			const exactMatches = providerIndex.get(consumerKey);
			if (exactMatches) {
				for (const p of exactMatches) {
					const segInfo = countSegmentTypes(normalizeHttpPath(c.address));
					candidates.push({
						providerFactUid: p.factUid,
						consumerFactUid: c.factUid,
						matchBasis: "address_match",
						confidence: computeHttpConfidence(c, segInfo),
					});
				}
				continue;
			}

			// Strategy 2: path-only match (consumer has unknown method or
			// method is wildcard "*"). Lower confidence.
			if (!method || method === "*") {
				const pathOnly = normalizeHttpPath(c.address);
				const pathMatches = providerByPathOnly.get(pathOnly);
				if (pathMatches) {
					for (const p of pathMatches) {
						const segInfo = countSegmentTypes(pathOnly);
						candidates.push({
							providerFactUid: p.factUid,
							consumerFactUid: c.factUid,
							matchBasis: "heuristic",
							confidence: computeHttpConfidence(c, segInfo) * 0.7,
						});
					}
				}
			}
		}

		return candidates;
	}
}

/**
 * Compute HTTP match confidence from consumer fact and path structure.
 *
 * Factors:
 *   - Consumer extraction confidence (literal > template, axios > fetch)
 *   - Path specificity: more literal segments = more discriminating match
 *   - All-literal paths get a slight boost (no wildcard ambiguity)
 *
 * This is a PROTOTYPE scoring model. Not calibrated against real-world
 * false-positive rates. Sufficient for model validation.
 */
function computeHttpConfidence(
	consumer: BoundaryConsumerFact,
	segInfo: { total: number; literal: number; wildcard: number },
): number {
	// Base: consumer's extraction confidence (0.7 - 0.95).
	let conf = consumer.confidence;

	// Path specificity factor:
	// Paths with more literal segments are more discriminating.
	// Single-segment paths (/health) are common but ambiguous.
	// Multi-segment API paths (/api/v2/products/{_}) are specific.
	if (segInfo.total <= 1) {
		conf *= 0.8; // Short paths are less discriminating.
	} else if (segInfo.wildcard === 0) {
		conf *= 1.0; // All-literal path — exact structural match.
	} else {
		// Mixed literal + wildcard: slight penalty per wildcard.
		const wildcardRatio = segInfo.wildcard / segInfo.total;
		conf *= 1.0 - wildcardRatio * 0.1;
	}

	// Clamp to [0, 1].
	return Math.max(0, Math.min(1, conf));
}

// ── Orchestrator ────────────────────────────────────────────────────

/** Default strategy registry. Extend as new mechanisms are added. */
const DEFAULT_STRATEGIES: BoundaryMatchStrategy[] = [
	new HttpBoundaryMatchStrategy(),
];

/**
 * Match boundary provider facts against consumer facts.
 *
 * Groups facts by mechanism, dispatches to the appropriate strategy,
 * and collects all link candidates. Facts with no registered strategy
 * are silently skipped (no match possible until a strategy is added).
 *
 * @param providers - All provider facts to match (must carry factUid). May span mechanisms.
 * @param consumers - All consumer facts to match (must carry factUid). May span mechanisms.
 * @param strategies - Override strategy registry (for testing). Defaults to all built-in strategies.
 */
export function matchBoundaryFacts(
	providers: MatchableProviderFact[],
	consumers: MatchableConsumerFact[],
	strategies: BoundaryMatchStrategy[] = DEFAULT_STRATEGIES,
): BoundaryLinkCandidate[] {
	// Build mechanism → strategy lookup.
	const strategyMap = new Map<BoundaryMechanism, BoundaryMatchStrategy>();
	for (const s of strategies) {
		strategyMap.set(s.mechanism, s);
	}

	// Group providers and consumers by mechanism.
	const providersByMechanism = new Map<BoundaryMechanism, MatchableProviderFact[]>();
	const consumersByMechanism = new Map<BoundaryMechanism, MatchableConsumerFact[]>();

	for (const p of providers) {
		const mech = p.mechanism;
		if (!providersByMechanism.has(mech)) providersByMechanism.set(mech, []);
		providersByMechanism.get(mech)!.push(p);
	}
	for (const c of consumers) {
		const mech = c.mechanism;
		if (!consumersByMechanism.has(mech)) consumersByMechanism.set(mech, []);
		consumersByMechanism.get(mech)!.push(c);
	}

	// Match within each mechanism.
	const allCandidates: BoundaryLinkCandidate[] = [];

	for (const [mech, strategy] of strategyMap) {
		const mechProviders = providersByMechanism.get(mech);
		const mechConsumers = consumersByMechanism.get(mech);
		if (!mechProviders || !mechConsumers) continue;

		const candidates = strategy.match(mechProviders, mechConsumers);
		allCandidates.push(...candidates);
	}

	return allCandidates;
}

// ── Exports for indexer use ─────────────────────────────────────────

/**
 * Get the match strategy for a mechanism.
 * Returns null if no strategy is registered.
 * Used by the indexer to compute matcher_key at persist time.
 */
export function getMatchStrategy(
	mechanism: BoundaryMechanism,
	strategies: BoundaryMatchStrategy[] = DEFAULT_STRATEGIES,
): BoundaryMatchStrategy | null {
	return strategies.find((s) => s.mechanism === mechanism) ?? null;
}

File: rust/crates/classification/src/boundary_matcher.rs
//! Boundary matcher — strategy-dispatched provider/consumer pairing.
//!
//! Mirror of `src/core/classification/boundary-matcher.ts`.
//!
//! Pure core business logic. No I/O, no storage, no framework
//! awareness. The orchestrator dispatches by mechanism to the
//! correct match logic.
//!
//! Architecture: the TS version uses a class-per-mechanism strategy
//! pattern (BoundaryMatchStrategy interface + HttpBoundaryMatchStrategy
//! + CliBoundaryMatchStrategy classes). The Rust port uses a simpler
//! enum-dispatch via `match` on `BoundaryMechanism` — more idiomatic
//! for a small fixed set of strategies and avoids trait objects.
//!
//! Maturity: PROTOTYPE. Same limitations as the TS version:
//! no query parameter matching, no route prefix/wildcard segment
//! matching, no cross-mechanism matching, confidence scoring is
//! a simple product (not calibrated).

use std::collections::HashMap;

use serde_json::{Map, Value};

use crate::types::{
	BoundaryLinkCandidate, BoundaryMatchBasis, BoundaryMechanism,
	MatchableConsumerFact, MatchableProviderFact,
};

/// Compute the normalized matcher key for a boundary fact.
///
/// Used at two sites:
///   - Persist time (indexer computes key for the DB column)
///   - Match time (strategy normalizes before comparing)
///
/// Returns `None` if the mechanism has no registered strategy.
///
/// Mirror of the TS `getMatchStrategy(mechanism)?.computeMatcherKey(address, metadata)`
/// pattern from `boundary-matcher.ts:432`.
///
/// This function is re-exported as `pub` from the crate root.
pub fn compute_matcher_key(
	mechanism: BoundaryMechanism,
	address: &str,
	metadata: &Map<String, Value>,
) -> Option<String> {
	match mechanism {
		BoundaryMechanism::Http => Some(http_compute_matcher_key(address, metadata)),
		BoundaryMechanism::CliCommand => Some(cli_compute_matcher_key(address)),
		_ => None,
	}
}

/// Match boundary provider facts against consumer facts.
///
/// Groups facts by mechanism, dispatches to the appropriate
/// match logic, and collects all link candidates. Facts with no
/// registered strategy are silently skipped.
///
/// Mirror of `matchBoundaryFacts` from `boundary-matcher.ts:384`.
///
/// This function is re-exported as `pub` from the crate root.
pub fn match_boundary_facts(
	providers: &[MatchableProviderFact],
	consumers: &[MatchableConsumerFact],
) -> Vec<BoundaryLinkCandidate> {
	// Group by mechanism.
	let mut providers_by_mech: HashMap<BoundaryMechanism, Vec<&MatchableProviderFact>> =
		HashMap::new();
	let mut consumers_by_mech: HashMap<BoundaryMechanism, Vec<&MatchableConsumerFact>> =
		HashMap::new();
	for p in providers {
		providers_by_mech
			.entry(p.fact.mechanism)
			.or_default()
			.push(p);
	}
	for c in consumers {
		consumers_by_mech
			.entry(c.fact.mechanism)
			.or_default()
			.push(c);
	}

	let mut all_candidates = Vec::new();

	// Dispatch per mechanism.
	for mech in [BoundaryMechanism::Http, BoundaryMechanism::CliCommand] {
		let Some(mech_providers) = providers_by_mech.get(&mech) else {
			continue;
		};
		let Some(mech_consumers) = consumers_by_mech.get(&mech) else {
			continue;
		};
		let candidates = match mech {
			BoundaryMechanism::Http => {
				http_match(mech_providers, mech_consumers)
			}
			BoundaryMechanism::CliCommand => {
				cli_match(mech_providers, mech_consumers)
			}
			_ => Vec::new(),
		};
		all_candidates.extend(candidates);
	}

	all_candidates
}

// ── HTTP match strategy ──────────────────────────────────────────

const PARAM_WILDCARD: &str = "{_}";

/// Normalize a single path segment. `{id}`, `{param}` → `{_}`;
/// `:id`, `:productId` → `{_}`; everything else → literal.
fn normalize_segment(segment: &str) -> &str {
	if segment.starts_with('{') && segment.ends_with('}') {
		return PARAM_WILDCARD;
	}
	if segment.starts_with(':') && segment.len() > 1 {
		return PARAM_WILDCARD;
	}
	segment
}

/// Normalize an HTTP path into a canonical segment representation.
fn normalize_http_path(path: &str) -> String {
	if path.is_empty() || path == "/" {
		return "/".to_string();
	}
	path.split('/')
		.map(normalize_segment)
		.collect::<Vec<_>>()
		.join("/")
}

struct SegmentInfo {
	total: usize,
	#[allow(dead_code)]
	literal: usize,
	wildcard: usize,
}

fn count_segment_types(normalized_path: &str) -> SegmentInfo {
	let segments: Vec<&str> = normalized_path
		.split('/')
		.filter(|s| !s.is_empty())
		.collect();
	let wildcard = segments.iter().filter(|&&s| s == PARAM_WILDCARD).count();
	SegmentInfo {
		total: segments.len(),
		literal: segments.len() - wildcard,
		wildcard,
	}
}

fn http_compute_matcher_key(address: &str, metadata: &Map<String, Value>) -> String {
	let method = metadata
		.get("httpMethod")
		.and_then(|v| v.as_str())
		.map(|m| m.to_uppercase())
		.unwrap_or_else(|| "*".to_string());
	let normalized_path = normalize_http_path(address);
	format!("{} {}", method, normalized_path)
}

fn http_match(
	providers: &[&MatchableProviderFact],
	consumers: &[&MatchableConsumerFact],
) -> Vec<BoundaryLinkCandidate> {
	if providers.is_empty() || consumers.is_empty() {
		return Vec::new();
	}

	// Build provider index: matcher_key → provider facts.
	let mut provider_index: HashMap<String, Vec<&MatchableProviderFact>> = HashMap::new();
	let mut provider_by_path_only: HashMap<String, Vec<&MatchableProviderFact>> =
		HashMap::new();

	for p in providers {
		let key = http_compute_matcher_key(&p.fact.address, &p.fact.metadata);
		provider_index.entry(key).or_default().push(p);

		let path_only = normalize_http_path(&p.fact.address);
		provider_by_path_only.entry(path_only).or_default().push(p);
	}

	let mut candidates = Vec::new();

	for c in consumers {
		let consumer_key = http_compute_matcher_key(&c.fact.address, &c.fact.metadata);
		let method = c
			.fact
			.metadata
			.get("httpMethod")
			.and_then(|v| v.as_str())
			.map(|m| m.to_uppercase());

		// Strategy 1: exact key match.
		if let Some(exact_matches) = provider_index.get(&consumer_key) {
			let seg_info = count_segment_types(&normalize_http_path(&c.fact.address));
			for p in exact_matches {
				candidates.push(BoundaryLinkCandidate {
					provider_fact_uid: p.fact_uid.clone(),
					consumer_fact_uid: c.fact_uid.clone(),
					match_basis: BoundaryMatchBasis::AddressMatch,
					confidence: compute_http_confidence(c.fact.confidence, &seg_info),
				});
			}
			continue;
		}

		// Strategy 2: path-only match (consumer has unknown method).
		match method.as_deref() {
			None | Some("*") => {
				let path_only = normalize_http_path(&c.fact.address);
				if let Some(path_matches) = provider_by_path_only.get(&path_only) {
					let seg_info = count_segment_types(&path_only);
					for p in path_matches {
						candidates.push(BoundaryLinkCandidate {
							provider_fact_uid: p.fact_uid.clone(),
							consumer_fact_uid: c.fact_uid.clone(),
							match_basis: BoundaryMatchBasis::Heuristic,
							confidence: compute_http_confidence(
								c.fact.confidence,
								&seg_info,
							) * 0.7,
						});
					}
				}
			}
			_ => {}
		}
	}

	candidates
}

fn compute_http_confidence(consumer_confidence: f64, seg_info: &SegmentInfo) -> f64 {
	let mut conf = consumer_confidence;
	if seg_info.total <= 1 {
		conf *= 0.8;
	} else if seg_info.wildcard == 0 {
		conf *= 1.0;
	} else {
		let wildcard_ratio = seg_info.wildcard as f64 / seg_info.total as f64;
		conf *= 1.0 - wildcard_ratio * 0.1;
	}
	conf.clamp(0.0, 1.0)
}

// ── CLI command match strategy ───────────────────────────────────

fn cli_compute_matcher_key(address: &str) -> String {
	address.to_lowercase().trim().to_string()
}

fn cli_match(
	providers: &[&MatchableProviderFact],
	consumers: &[&MatchableConsumerFact],
) -> Vec<BoundaryLinkCandidate> {
	if providers.is_empty() || consumers.is_empty() {
		return Vec::new();
	}

	let mut provider_index: HashMap<String, Vec<&MatchableProviderFact>> = HashMap::new();
	for p in providers {
		let key = cli_compute_matcher_key(&p.fact.address);
		provider_index.entry(key).or_default().push(p);
	}

	let mut candidates = Vec::new();

	for c in consumers {
		let consumer_key = cli_compute_matcher_key(&c.fact.address);

		// Strategy 1: exact command path match.
		if let Some(exact_matches) = provider_index.get(&consumer_key) {
			for p in exact_matches {
				candidates.push(BoundaryLinkCandidate {
					provider_fact_uid: p.fact_uid.clone(),
					consumer_fact_uid: c.fact_uid.clone(),
					match_basis: BoundaryMatchBasis::OperationMatch,
					confidence: c.fact.confidence * 0.9,
				});
			}
			continue;
		}

		// Strategy 2: binary-prefix strip.
		// Consumer paths from scripts include the binary name
		// ("rgr repo add"), but providers register only the
		// subcommand path ("repo add"). Guard: stripped remainder
		// must have >= 2 tokens to avoid ambiguous single-word
		// matches.
		let tokens: Vec<&str> = consumer_key.split_whitespace().collect();
		if tokens.len() >= 3 {
			let without_binary = tokens[1..].join(" ");
			if let Some(prefix_matches) = provider_index.get(&without_binary) {
				for p in prefix_matches {
					candidates.push(BoundaryLinkCandidate {
						provider_fact_uid: p.fact_uid.clone(),
						consumer_fact_uid: c.fact_uid.clone(),
						match_basis: BoundaryMatchBasis::Heuristic,
						confidence: c.fact.confidence * 0.75,
					});
				}
			}
		}
	}

	candidates
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::types::{
		BoundaryConsumerBasis, BoundaryConsumerFact, BoundaryProviderBasis,
		BoundaryProviderFact,
	};
	use serde_json::json;

	fn make_provider(
		fact_uid: &str,
		mechanism: BoundaryMechanism,
		address: &str,
		metadata: Map<String, Value>,
	) -> MatchableProviderFact {
		MatchableProviderFact {
			fact_uid: fact_uid.to_string(),
			fact: BoundaryProviderFact {
				mechanism,
				operation: format!("GET {}", address),
				address: address.to_string(),
				handler_stable_key: "handler".to_string(),
				source_file: "src/app.ts".to_string(),
				line_start: 10,
				framework: "express".to_string(),
				basis: BoundaryProviderBasis::Registration,
				schema_ref: None,
				metadata,
			},
		}
	}

	fn make_consumer(
		fact_uid: &str,
		mechanism: BoundaryMechanism,
		address: &str,
		confidence: f64,
		metadata: Map<String, Value>,
	) -> MatchableConsumerFact {
		MatchableConsumerFact {
			fact_uid: fact_uid.to_string(),
			fact: BoundaryConsumerFact {
				mechanism,
				operation: format!("GET {}", address),
				address: address.to_string(),
				caller_stable_key: "caller".to_string(),
				source_file: "src/client.ts".to_string(),
				line_start: 20,
				basis: BoundaryConsumerBasis::Literal,
				confidence,
				schema_ref: None,
				metadata,
			},
		}
	}

	fn http_meta(method: &str) -> Map<String, Value> {
		let mut m = Map::new();
		m.insert("httpMethod".to_string(), json!(method));
		m
	}

	// ── computeMatcherKey (HTTP) ─────────────────────────────

	#[test]
	fn http_key_normalizes_spring_param() {
		let key = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api/orders/{id}",
			&http_meta("GET"),
		);
		assert_eq!(key.unwrap(), "GET /api/orders/{_}");
	}

	#[test]
	fn http_key_normalizes_express_param() {
		let key = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api/orders/:id",
			&http_meta("POST"),
		);
		assert_eq!(key.unwrap(), "POST /api/orders/{_}");
	}

	#[test]
	fn http_key_preserves_literal_segments() {
		let key = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api/v2/products",
			&http_meta("GET"),
		);
		assert_eq!(key.unwrap(), "GET /api/v2/products");
	}

	#[test]
	fn http_key_uses_star_for_unknown_method() {
		let key = compute_matcher_key(
			BoundaryMechanism::Http,
			"/health",
			&Map::new(),
		);
		assert_eq!(key.unwrap(), "* /health");
	}

	#[test]
	fn http_key_uppercases_method() {
		let key = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api",
			&http_meta("get"),
		);
		assert_eq!(key.unwrap(), "GET /api");
	}

	#[test]
	fn spring_and_express_params_produce_same_key() {
		let spring = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api/v2/products/{id}",
			&http_meta("GET"),
		);
		let express = compute_matcher_key(
			BoundaryMechanism::Http,
			"/api/v2/products/:productId",
			&http_meta("GET"),
		);
		assert_eq!(spring, express);
	}

	// ── matchBoundaryFacts (HTTP) ────────────────────────────

	#[test]
	fn http_matches_provider_param_to_consumer_param() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::Http,
			"/api/orders/{id}",
			http_meta("GET"),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::Http,
			"/api/orders/{param}",
			0.9,
			http_meta("GET"),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert_eq!(links.len(), 1);
		assert_eq!(links[0].provider_fact_uid, "p1");
		assert_eq!(links[0].consumer_fact_uid, "c1");
		assert_eq!(links[0].match_basis, BoundaryMatchBasis::AddressMatch);
	}

	#[test]
	fn http_does_not_match_different_methods() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::Http,
			"/api/orders",
			http_meta("GET"),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::Http,
			"/api/orders",
			0.9,
			http_meta("POST"),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert!(links.is_empty());
	}

	#[test]
	fn http_does_not_match_across_mechanisms() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::Http,
			"/api/orders",
			http_meta("GET"),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::Grpc,
			"/api/orders",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert!(links.is_empty());
	}

	#[test]
	fn http_path_only_fallback_for_unknown_method() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::Http,
			"/api/orders",
			http_meta("GET"),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::Http,
			"/api/orders",
			0.9,
			Map::new(), // no httpMethod → path-only fallback
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert_eq!(links.len(), 1);
		assert_eq!(links[0].match_basis, BoundaryMatchBasis::Heuristic);
		assert!(
			links[0].confidence < 0.9 * 0.8, // heuristic penalty
			"path-only confidence must be lower than exact"
		);
	}

	// ── Confidence ───────────────────────────────────────────

	#[test]
	fn literal_consumer_has_higher_confidence_than_wildcard() {
		let seg_literal = count_segment_types("/api/v2/products");
		let seg_wildcard = count_segment_types("/api/v2/products/{_}");
		let conf_literal = compute_http_confidence(0.9, &seg_literal);
		let conf_wildcard = compute_http_confidence(0.9, &seg_wildcard);
		assert!(conf_literal > conf_wildcard);
	}

	#[test]
	fn short_path_has_lower_confidence() {
		let seg_short = count_segment_types("/health");
		let seg_long = count_segment_types("/api/v2/products");
		let conf_short = compute_http_confidence(0.9, &seg_short);
		let conf_long = compute_http_confidence(0.9, &seg_long);
		assert!(conf_short < conf_long);
	}

	// ── computeMatcherKey (CLI) ──────────────────────────────

	#[test]
	fn cli_key_lowercases_and_trims() {
		let key = compute_matcher_key(
			BoundaryMechanism::CliCommand,
			"  Repo Add  ",
			&Map::new(),
		);
		assert_eq!(key.unwrap(), "repo add");
	}

	// ── matchBoundaryFacts (CLI) ─────────────────────────────

	#[test]
	fn cli_exact_match_produces_operation_match_basis() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::CliCommand,
			"repo add",
			Map::new(),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::CliCommand,
			"repo add",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert_eq!(links.len(), 1);
		assert_eq!(links[0].match_basis, BoundaryMatchBasis::OperationMatch);
	}

	#[test]
	fn cli_binary_prefix_strip_matches_with_heuristic_basis() {
		// Consumer "rgr repo add" → strip binary "rgr" → "repo add"
		// matches provider "repo add". Guard: 3 tokens ≥ 3, so the
		// stripped remainder ("repo add") has 2 tokens which is OK.
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::CliCommand,
			"repo add",
			Map::new(),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::CliCommand,
			"rgr repo add",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert_eq!(links.len(), 1);
		assert_eq!(links[0].match_basis, BoundaryMatchBasis::Heuristic);
		assert!(
			links[0].confidence < 0.9 * 0.9,
			"binary-prefix heuristic must have lower confidence than exact"
		);
	}

	#[test]
	fn cli_binary_prefix_strip_requires_at_least_3_tokens() {
		// Consumer "cargo test" has only 2 tokens. Stripping the
		// binary ("cargo") leaves a single token ("test"), which
		// is too ambiguous. The guard (tokens.len() >= 3) prevents
		// false positives where "cargo test" would match a provider
		// named just "test".
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::CliCommand,
			"test",
			Map::new(),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::CliCommand,
			"cargo test",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert!(
			links.is_empty(),
			"2-token consumer must NOT binary-prefix-strip to a single-word provider"
		);
	}

	#[test]
	fn cli_exact_match_is_case_insensitive() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::CliCommand,
			"Repo Add",
			Map::new(),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::CliCommand,
			"repo add",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert_eq!(links.len(), 1, "CLI matching should be case-insensitive");
	}

	#[test]
	fn cli_no_match_for_different_commands() {
		let providers = vec![make_provider(
			"p1",
			BoundaryMechanism::CliCommand,
			"repo add",
			Map::new(),
		)];
		let consumers = vec![make_consumer(
			"c1",
			BoundaryMechanism::CliCommand,
			"repo remove",
			0.9,
			Map::new(),
		)];
		let links = match_boundary_facts(&providers, &consumers);
		assert!(links.is_empty());
	}

	// ── Unknown mechanism ────────────────────────────────────

	#[test]
	fn unknown_mechanism_returns_none_for_matcher_key() {
		let key = compute_matcher_key(
			BoundaryMechanism::Grpc,
			"/some.Service/Method",
			&Map::new(),
		);
		assert!(key.is_none());
	}
}
